azure/src/main/java/org/apache/iceberg/azure/adlsv2/ADLSOutputStream.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.azure.adlsv2;
+
+import com.azure.storage.common.ParallelTransferOptions;
+import com.azure.storage.file.datalake.DataLakeFileClient;
+import com.azure.storage.file.datalake.options.DataLakeFileOutputStreamOptions;
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+import org.apache.iceberg.azure.AzureProperties;
+import org.apache.iceberg.io.FileIOMetricsContext;
+import org.apache.iceberg.io.PositionOutputStream;
+import org.apache.iceberg.metrics.Counter;
+import org.apache.iceberg.metrics.MetricsContext;
+import org.apache.iceberg.metrics.MetricsContext.Unit;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ADLSOutputStream extends PositionOutputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(ADLSOutputStream.class);
+
+  private final StackTraceElement[] createStack;
+  private final DataLakeFileClient fileClient;
+  private final AzureProperties azureProperties;
+
+  private OutputStream stream;
+
+  private final Counter writeBytes;
+  private final Counter writeOperations;
+
+  private long pos = 0;
+  private boolean closed = false;
+
+  ADLSOutputStream(
+      DataLakeFileClient fileClient, AzureProperties azureProperties, MetricsContext metrics)
+      throws IOException {
+    this.fileClient = fileClient;
+    this.azureProperties = azureProperties;
+
+    createStack = Thread.currentThread().getStackTrace();
```

Comment:
nit: assignment should be preceded by `this.`

Reply:
Looks like we haven't been doing a great job of this, but the general convention is member assignment should always be preceded by `this.` (not just in the constructor).

Reply:
I went ahead and made this change. Though it can be a little bit inconsistent, for example, in the following sample, we're using a mixture of references which has me wondering why only use it for the assignment:
```
    if (closed) {
      return;
    }
    super.close();
    this.closed = true;
```